packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -1824,9 +1824,106 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
   return root;
 }
 
-export function warnIfNotCurrentlyBatchingInDev(fiber: Fiber): void {
+// in a test-like environment, we want to warn if dispatchAction() is
+// called outside of a TestUtils.act(...)/batchedUpdates/render call.
+// so we have a a step counter for when we descend/ascend from
+// actedUpdates() calls, and test on it for when to warn
+let actingUpdatesScopeDepth = 0;
+
+// we 'wait' for queued microtasks to resolve (after flushing effects)
+// using either setImmediate when available, or postMessage
+const nextTick =
+  typeof setImmediate !== 'undefined'
+    ? setImmediate
+    : callback => {
+        const channel = new MessageChannel();
```

Comment:
What if `MessageChannel` is not defined?

Reply:
so this is in older browsers (ie9 and below and some others). the annoying scenario is when someone's using fakeTimers + async act, and running their tests across browser. we can't use setTimeout in this case because it would require manually advancing after each act call. So the challenge is to delay by one macrotask, but without using any of the usual suspects - setTimeout, setImmediate, etc. 

Here's a super dirty version of a workaround
```jsx
function nextTick(callback){
  let promise = Promise.resolve()
  let ctr = 0
  while(ctr < 100){
    ctr++
    promise = promise.then(() => {})
  }
  return promise.then(() => {
    callback()
  })
};
```
This isn't satisfying; it's slow, and there are obvious edge cases. 

Another option is to create a dummy element (like an img/iframe/link) and try to load an inlined resource. something like this -
```jsx
function nextTick(callback){
  const img = new Image()
  img.onload = callback
  img.src = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII='
}
```
this one's less ickier. I tested it, and it "works". Thoughts?

Reply:
I don't think we care that TestUtils work in IE9.

Reply:
so... should I not have the fallback, or just throw an error? our 'support' usually involves just adding polyfills, and this is a scenario where that might be enough (it'll be odd to ask folks to install a MessageChannel polyfill to pass their tests, but sure.)

Reply:
If we can polyfill it easily (and the polyfill lives outside production code path) then it seems ok. Otherwise — I'm not sure it's reasonable that everyone using ReactDOM in prod pays the cost for the hypothetical users who run TestUtils in IE9.

Reply:
I'm moving all this logic into TestUtils, so it won't add to ReactDOM